6599 unable to set prescribed quantity when no stock

GeronimoJohn opened the pull request:
<!-- IMPORTANT!
  - Every PR must reference an issue; this helps to explain the intent of the PR
 -->

Fixes #6599

# 👩🏻‍💻 What does this PR do?

We want to enable saving an item within a prescription even when there is no stock, provided that the user has set a prescribed quantity.

To achieve this, we have introduced a new mutation called `set_prescribed_quantity` within the `batch_prescription` mutation. This focuses on saving an item despite the absence of stock.

<!-- Explain the changes you made -->

<!-- why are the changes needed -->

<!-- Add a screenshot if there are UI changes  -->

## 💌 Any notes for the reviewer?

<!-- Do you have any specific questions for the reviewer? -->

<!-- Is there a high risk/complicated change they should focus on? -->

<!-- any general areas of the codebase touched? any side effects caused? -->

<!-- Anything half cooked but going to be finished off in a different PR? -->

# 🧪 Testing

<!-- Explain the steps you'd take to test the changes of this PR manually -->

- [ ] Navigate to any prescriptions
- [ ] Add a new item with no stock
- [ ] Enter a prescribed quantity to the selected item
- [ ] Save button should get enabled - click on the save button
- [ ] Item should get saved!

# 📃 Documentation

- [ ] **Part of an epic**: documentation will be completed for the feature as a whole
- [ ] **No documentation required**: no user facing changes or a bug fix which isn't a change in behaviour
- [ ] **These areas should be updated or checked**: <!-- _(e.g.)_ New `issued` column in `Requisitions` indicates stock quantity already in shipments -->
  1.
  2.


# 📃 Reviewer Checklist

The PR Reviewer(s) should fill out this section before approving the PR

**Breaking Changes**
- [ ] No Breaking Changes in the Graphql API
- [ ] Technically some Breaking Changes but not expected to impact any integrations

**Issue Review**
- [ ] All requirements in original issue have been covered
- [ ] A follow up issue(s) have been created to cover additional requirements

**Tests Pass**
- [ ] Postgres
- [ ] SQLite
- [ ] Frontend

GeronimoJohn commented:
Found a few issues - working on fix

jmbrunskill commented:
![image](https://github.com/user-attachments/assets/625a6058-f7c5-47cd-888c-0f8bd161dc54)

jmbrunskill commented:
![image](https://github.com/user-attachments/assets/5ffbe5f9-fdc4-4382-b54b-fc629b0f0c4a)

jmbrunskill commented:
Not a major but something like this would be nice...
![image](https://github.com/user-attachments/assets/12fcf73b-0aa0-4403-972b-8e9aed7344c2)

jmbrunskill commented:
After adding stock, the prescribed quantity disappears?
![image](https://github.com/user-attachments/assets/a6000bc7-1640-42ac-bd2f-07133e9bdd01)

(Not a major btw, I think this is an edge case)

jmbrunskill commented:
> After adding stock, the prescribed quantity disappears? 

Actually this seems to be a general problem...
![image](https://github.com/user-attachments/assets/24bfb46e-3f02-4f5a-a123-7e6204169aad)

jmbrunskill reviewed (changes requested):
Can you remove prescribed quantity from 

`InsertPrescriptionLineInput` and `UpdatePrescriptionLineInput` please?

I think it will be confusing having two ways to update the same thing...

jmbrunskill commented on server/service/src/invoice_line/stock_out_line/set_prescribed_quantity/mod.rs:
```diff
@@ -0,0 +1,338 @@
+                    .invoice_id(EqualFilter::equal_to(&input.invoice_id)),
+            )?;
+
+            if existing_lines.len() > 1 {
```
I think this logic should have some comments to explain the intended behaviour.

E.g.

When we set prescribed quantity for an item, there's 3 scenarios.

1. We don't have any stock lines already create for that item, in which case we need to create an unallocated line for the prescription to record the prescribed quantity against
2. We already have more than one stock line, we need to update the one that already has prescribed quantity, or assign it to a single line if none of them have it.
3. If there's a proper allocated line, and an unallocated one, we need to remove the unallocated one just to clean up

jmbrunskill commented on server/service/src/invoice_line/stock_out_line/set_prescribed_quantity/mod.rs:
```diff
@@ -0,0 +1,338 @@
+        assert!(result.is_ok());
+
+        let invoice_line = result.unwrap();
+
```
I think you need a test here to see how many lines exist after this function runs. 
We should be deleting the unallocated line right?

jmbrunskill replied:
Would also like a test where there's more than one allocated stockline for the item.
We should check that there's only one line with prescribed quantity set on it in this scenario.

jmbrunskill commented on server/service/src/invoice_line/stock_out_line/set_prescribed_quantity/mod.rs:
```diff
@@ -0,0 +1,338 @@
+                .iter()
+                .find(|line| line.invoice_line_row.stock_line_id.is_some());
+
+            let new_line;
+            if let Some(existing_line) = existing_line_with_stock {
+                let mut updated_line = existing_line.clone();
+                updated_line.invoice_line_row.prescribed_quantity = Some(input.prescribed_quantity);
+                invoice_line_row_repo.upsert_one(&updated_line.invoice_line_row)?;
+                new_line = updated_line.invoice_line_row
+            } else {
+                let invoice_line = generate(uuid(), item_row, input.clone())?;
+                invoice_line_row_repo.upsert_one(&invoice_line)?;
+                new_line = invoice_line;
+            }
```
```suggestion
            let new_line = match existing_line_with_stock {
                Some(existing_line) => {
                    let mut updated_line = existing_line.clone();
                    updated_line.invoice_line_row.prescribed_quantity =
                        Some(input.prescribed_quantity);
                    invoice_line_row_repo.upsert_one(&updated_line.invoice_line_row)?;
                    updated_line.invoice_line_row
                }
                None => {
                    let invoice_line = generate(uuid(), item_row, input.clone())?;
                    invoice_line_row_repo.upsert_one(&invoice_line)?;
                    invoice_line
                }
            };
```

Using a match here is more `Rusty` than the if else.
I'm actually quite surprised your code worked without assigning `new_line` as `mut`?

jmbrunskill reviewed (commented):
Have you tried this with Edit prescribedQuantity turned off?

<img width="916" alt="image" src="https://github.com/user-attachments/assets/7c8ba5e7-47c3-4f64-984c-801af4bacf02" />

jmbrunskill commented:
> Have you tried this with Edit prescribedQuantity turned off?
> 

When I do I get this...

![image](https://github.com/user-attachments/assets/eff52cec-2f28-4373-a211-eb93fba12a46)

I've also noticed that we're sending a lot of duplicate setPrescribed Quantity requests???

![image](https://github.com/user-attachments/assets/bd469ac8-3f77-4be8-ad86-020f6c678398)


We should only send 1 per item...

And Ideally when the store pref is off, we wouldn't send any at all

GeronimoJohn commented:
Thanks for the review @jmbrunskill 

Clean up `set_prescribed_quantity` changes: https://github.com/msupply-foundation/open-msupply/pull/6883/commits/539c90face0f9a7e1167bd19e0a287675c976530

Remove `prescribed_quantity` from `InsertPrescriptionLineInput` and `UpdatePrescriptionLineInput`: https://github.com/msupply-foundation/open-msupply/pull/6883/commits/c5dbe7b0e8aacd072ad4aa5965c6bc1558dbc4f5

NaN issue fix: https://github.com/msupply-foundation/open-msupply/pull/6883/commits/b922c5dbf086d52d2e1aec7092df60516d877fab

GeronimoJohn commented:
> Have you tried this with Edit prescribedQuantity turned off?
> 
> <img alt="image" width="916" src="https://private-user-images.githubusercontent.com/8287373/421108143-7c8ba5e7-47c3-4f64-984c-801af4bacf02.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2NTc2NzksIm5iZiI6MTc0MTY1NzM3OSwicGF0aCI6Ii84Mjg3MzczLzQyMTEwODE0My03YzhiYTVlNy00N2MzLTRmNjQtOTg0Yy04MDFhZjRiYWNmMDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMDE0MjU5WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZGU2MTU5MzQ5ZjFkMGIyMWUyZTBkYzBkZmZmMzY4OTc0OWQ3M2U4N2Q3OGI5MGI5Mzg4NGUyYmRmYjliNGVlYyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.O5Af4kQYo6pjiwuaFa6rvcwd87FafkZ3RMzNCEH9anc">

Yup! Prescribed Quantity input gets hidden, or at least it should 😁 

![Screenshot 2025-03-11 at 2 48 16 PM](https://github.com/user-attachments/assets/9e924bf6-ee20-4cec-a7fc-c25e697cab2c)

GeronimoJohn commented:
> ![image](https://private-user-images.githubusercontent.com/8287373/421101756-5ffbe5f9-fdc4-4382-b54b-fc629b0f0c4a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2NTc2NzgsIm5iZiI6MTc0MTY1NzM3OCwicGF0aCI6Ii84Mjg3MzczLzQyMTEwMTc1Ni01ZmZiZTVmOS1mZGM0LTQzODItYjU0Yi1mYzYyOWIwZjBjNGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMDE0MjU4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjdkMTNmYzNkY2QxYzM0Y2FjOTY5ZDdmNGYxYWIyYjk3OWU4MGIyYzk4NmMwMDM3YjY2ZTM2MWJlZTZmYzcyYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.nM9pZ25-XE3CpNSTa_vgqwhuWK9rm5cf2blCYzS2_Oc)

What would be the expected behaviour here? In my head it kinda sense to not change status since the items are just placeholders

GeronimoJohn commented:
> After adding stock, the prescribed quantity disappears? ![image](https://private-user-images.githubusercontent.com/8287373/421102521-a6000bc7-1640-42ac-bd2f-07133e9bdd01.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDE2NTkzNzcsIm5iZiI6MTc0MTY1OTA3NywicGF0aCI6Ii84Mjg3MzczLzQyMTEwMjUyMS1hNjAwMGJjNy0xNjQwLTQyYWMtYmQyZi0wNzEzM2U5YmRkMDEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDMxMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAzMTFUMDIxMTE3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Mzc0YWM4MDc4ZmEyZTUxMTk5Y2Y1NGVmYjIyMzE4MTY4NWNhMjQwNWVmYmM3MTg1OWYyZjk1NTg4MjMxMmUwNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.EmRE3n7tJE_47m34DIeZKB2mNteovdrjQCwAzu_2yTA)
> 
> (Not a major btw, I think this is an edge case)

Fixed the issue within: https://github.com/msupply-foundation/open-msupply/commit/be9fbc0590674e47d976fe7e20c6a3dd4dadf0b5 (I think 😄)

https://github.com/user-attachments/assets/c0829fc4-c8ab-4e79-bbd3-a8359384e44f

jmbrunskill reviewed (approved):
Thanks @GeronimoJohn working well from what I can tell!

Note: I've updated your tests cases, it might be good if you can review that commit before merging.

There were a few problems in the tests and I thought I'd just do a comit rather than suggestion branches explainations etc.

1. You had duplicate ids for the invoice line row, this meant that the 3 lines was actually only 1 due to upsert logic
2.  Updated the tests to query from database rather than rely on mocks which aren't actually updated by the logic
3. Added tested to check the unallocated line is deleted.

GeronimoJohn commented:
> Thanks @GeronimoJohn working well from what I can tell!
> 
> Note: I've updated your tests cases, it might be good if you can review that commit before merging.
> 
> There were a few problems in the tests and I thought I'd just do a comit rather than suggestion branches explainations etc.
> 
> 1. You had duplicate ids for the invoice line row, this meant that the 3 lines was actually only 1 due to upsert logic
> 2. Updated the tests to query from database rather than rely on mocks which aren't actually updated by the logic
> 3. Added tested to check the unallocated line is deleted.
> 
> [f80676f](https://github.com/msupply-foundation/open-msupply/commit/f80676f1394aa3e664c1da9932f6781e5a1013c4)

Thanks @jmbrunskill! I've reviewed the commit, and that makes sense. I'm kinda facepalming a bit over points 1 and 2—didn't catch that while writing it 😆
